### PR TITLE
gh-127330: Update for OpenSSL 3.4

### PR DESCRIPTION
**Pull request by encukou**
## Add `git describe` output to headers generated by `make_ssl_data.py`

IMO, this info is more important than the date when the file was generated.
It does mean that the tool now requires a Git checkout of OpenSSL, not for example a release tarball.

I've regenerated the older file to add the info. To the other older file I added a note about manual edits.

## Add notes on how to add a new OpenSSL version

I believe that what I wrote is the correct process. [ssl experts](https://devguide.python.org/core-developers/experts/), do you want to you verify? @jackjansen, @tiran, @dstufft, @alex

## Add 3.4 error messages and multissl tests

Hopefully we'll start getting fewer `[SYS] unknown error` in tests! (See #127257)

* Issue: gh-127330

### CODE REVIEW

**Comment by encukou** [comment]
!buildbot Arch

**Comment by alex** [comment]
I think this looks right, but I've never actually done the process myself :-)

The only other thing we might want is bumping the version that's used in the Windows and macOS builds.

**Comment by ned-deily** [comment]
I believe @gpshead has done these updates most recently.

**Comment by petermarko** [comment]
This does not seem to fix #125936
After picking commit "Generate data for OpenSSL 3.4 and add it to multissltests" to 3.13.0, the test are still failing.
I don't see the afected tests being run in https://github.com/python/cpython/actions/runs/12051565064/job/33602924793
Not sure what I'm missing here...

File: Tools/ssl/make_ssl_data.py [hunk 1]
```diff
@@ -117,9 +137,17 @@ def main():
     # sort by libname, numeric error code
     args.reasons = sorted(reasons, key=operator.itemgetter(0, 3))
 
+    git_describe = subprocess.run(
+        ['git', 'describe', '--long', '--dirty'],
+        cwd=args.srcdir,
+        capture_output=True,
+        encoding='latin-1',
```
Comment (gpshead) [inline Tools/ssl/make_ssl_data.py]: eek!

**Review by gpshead — APPROVED** [review]

File: Tools/ssl/make_ssl_data.py (hunk 1 above)
Comment (encukou) [reply Tools/ssl/make_ssl_data.py]:
```suggestion
        encoding='utf-8',
```

**Comment by encukou** [comment]
@petermarko, yes, this is better error reporting in general, but doesn't affect the bug we're getting there. I'll send another PR for a better error message for that case, and *then*, hopefully, one to fix the underlying bug.

File: Tools/ssl/make_ssl_data.py [hunk 2]
```diff
@@ -5,16 +5,36 @@
...
+- the path to the OpenSSL git checkout
 - the path to the header file to be generated Modules/_ssl_data_{version}.h
 - error codes are version specific
+
+The OpenSSL git checkout should be at a specific tag, using commands like:
+    git tag --list 'openssl-*'
+    git switch --detach openssl-3.4.0
+
+
+After generating the definitions, the result with newest pre-existing file.
```
Comment (encukou) [inline Tools/ssl/make_ssl_data.py]:
```suggestion
After generating the definitions, compare the result with newest pre-existing file.
```

**Comment by rossburton** [comment]
Any chance of this being backported to the 3.13 branch?

**Comment by encukou** [comment]
@Yhg1s Do you want Python 3.13 to have messages for OpenSSL 3.4's new error codes?
I see it as a feature but wouldn't mind doing the backport.

**Comment by encukou** [comment]
We now have PRs for OpenSSL 3.4.1 (#131618) and 3.5 (#137720). All error messages, plus improvements to the scripts that generate the files.

@Yhg1s, the question still stands. I'm happy to do the backports if you want them.

**Comment by Yhg1s** [comment]
Ah, sorry, I thought we'd discussed this (or maybe I just didn't make the final call). No, I don't think we should backport this to 3.13. It may seem harmless but it's the kind of thing that'll trip people up when doing patch upgrades (which we want to be as seamless as possible, evidence of the last couple of expedited releases notwithstanding.)

**Comment by encukou** [comment]
It's also possible I didn't *record* your call correctly back then.
Thanks for making it clear!

**Pull request merged**